### PR TITLE
interop-testing: Add status_code_and_message interop test

### PR DESCRIPTION
**Pull request by ericgribkoff**
This implements the [status_code_and_message](https://github.com/grpc/grpc/blob/f29de2619476622e61c894559f8885ca3cc8971c/doc/interop-test-descriptions.md#status_code_and_message)  interop test (#1993, https://github.com/grpc/grpc/issues/8481)

### CODE REVIEW

File: interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java [hunk 1]
```diff
@@ -843,6 +844,50 @@ public void gracefulShutdown() throws Exception {
...
+    if (metricsExpected()) {
+      assertClientMetrics("grpc.testing.TestService/UnaryCall", Status.Code.UNKNOWN);
+    }
+
+    // Test FullDuplexCall
+    @SuppressWarnings("unchecked")
+    StreamObserver<StreamingOutputCallResponse> responseObserver = mock(StreamObserver.class);
+    StreamObserver<StreamingOutputCallRequest> requestObserver
+        = asyncStub.fullDuplexCall(responseObserver);
+    requestObserver.onNext(streamingRequest);
```
Comment (ericgribkoff) [inline interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java]: The test spec requires the client to half-close here. Working on adding this now.

Comment (ericgribkoff) [inline interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java]: Done.

File: interop-testing/src/main/java/io/grpc/testing/integration/TestServiceImpl.java [hunk 2]
```diff
@@ -186,13 +195,22 @@ public void onError(Throwable cause) {
     return new StreamObserver<StreamingOutputCallRequest>() {
       @Override
```
Comment (dapengzhang0) [inline interop-testing/src/main/java/io/grpc/testing/integration/TestServiceImpl.java]: I would rather add the boolean field `canceled` in this anonymous `StreamObserver` class than in the `ResponseDispatcher` class. (add `ResponseDispatcher#clear()` if necessary)

File: interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java [hunk 3]
```diff
@@ -924,6 +925,51 @@ public void customMetadata() throws Exception {
     }
   }
 
+  @Test(timeout = 10000)
+  public void statusCodeAndMessage() throws Exception {
+    final int errorCode = 2;
```
Comment (ericgribkoff) [inline interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java]: Done

File: interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java [hunk 4]
```diff
@@ -924,6 +925,51 @@ public void customMetadata() throws Exception {
...
+      assertEquals(Status.UNKNOWN.getCode(), e.getStatus().getCode());
+      assertEquals(errorMessage, e.getStatus().getDescription());
+    }
+    if (metricsExpected()) {
+      assertClientMetrics("grpc.testing.TestService/UnaryCall", Status.Code.UNKNOWN);
+    }
+
+    // Test FullDuplexCall
+    @SuppressWarnings("unchecked")
+    StreamObserver<StreamingOutputCallResponse> responseObserver = mock(StreamObserver.class);
```
Comment (ericgribkoff) [inline interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java]: Done

File: interop-testing/src/main/java/io/grpc/testing/integration/TestServiceImpl.java (hunk 2 above)
Comment (ericgribkoff) [inline interop-testing/src/main/java/io/grpc/testing/integration/TestServiceImpl.java]: Why is the boolean in the `StreamObserver` preferred? I can do this, but it still requires some kind of check in `ResponseDispatcher#dispatchChunk()`, see below...

File: interop-testing/src/main/java/io/grpc/testing/integration/TestServiceImpl.java [hunk 5]
```diff
@@ -296,10 +315,28 @@ public ResponseDispatcher completeInput() {
...
+    public synchronized boolean isCancelled() {
+      return cancelled;
+    }
+
+    /**
      * Dispatches the current response chunk to the client. This is only called by the executor. At
      * any time, a given dispatch task should only be registered with the executor once.
      */
     private synchronized void dispatchChunk() {
+      if (cancelled) {
```
Comment (ericgribkoff) [inline interop-testing/src/main/java/io/grpc/testing/integration/TestServiceImpl.java]: If `canceled` is no longer part of the `ResponseDispatcher` state, all I can do here is check if the queue is empty, e.g.:

```
    private synchronized void dispatchChunk() {
      // If the service already called cancel(), the queue may be empty.
      if (chunks.isEmpty()) {
        return;
      }
      ...
```

It seems like this has the potential to mask other problems with the code by potentially masking cases where the queue is incorrectly empty at this stage. This doesn't seem to be possible with the current code, but still hesitant about adding this kind of check.

File: interop-testing/src/main/java/io/grpc/testing/integration/TestServiceImpl.java [hunk 6]
```diff
@@ -296,10 +315,28 @@ public ResponseDispatcher completeInput() {
     }
 
     /**
+     * Allows the service to cancel the remaining responses.
+     */
+    public synchronized void cancel() {
+      if (cancelled) {
```
Comment (dapengzhang0) [inline interop-testing/src/main/java/io/grpc/testing/integration/TestServiceImpl.java]: nit: `Precondition.checkState(!cancelled, "Dispatcher already cancelled");`

**Comment by ericgribkoff** [comment]
retest this please

**Pull request merged**